### PR TITLE
[IMP][website_mail] improved template designer, added support for backgr...

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
...ound color, font color and bootstrap buttons(inline css) in mail designer + added new snippets + improved snippet categories
- Task: https://www.odoo.com/web?#id=9439&view_type=form&model=project.task
- Pad: https://pad.odoo.com/p/r.7aiZsyw4go0bC7De

### CODE REVIEW

File: addons/mass_mailing/models/mail_mail.py [hunk 1]
```diff
@@ -88,12 +88,14 @@ def send_get_mail_body(self, cr, uid, mail, partner=None, context=None):
...
         res = super(MailMail, self).send_get_email_dict(cr, uid, mail, partner, context=context)
+        base_url = self.pool.get('ir.config_parameter').get_param(cr, uid, 'web.base.url')
         if mail.mailing_id and res.get('body') and res.get('email_to'):
             emails = tools.email_split(res.get('email_to')[0])
             email_to = emails and emails[0] or False
-            unsubscribe_url = self._get_unsubscribe_url(cr, uid, mail, email_to, context=context)
-            if unsubscribe_url:
-                res['body'] = tools.append_content_to_html(res['body'], unsubscribe_url, plaintext=False, container_tag='p')
+            unsubscribe_url, msg = self._get_unsubscribe_url(cr, uid, mail, email_to, context=context)
+            link_to_replace =  base_url+'/unsubscribe_from_list'
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mail_mail.py]: Why this hack ?

File: addons/mass_mailing/models/mail_mail.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ def _get_unsubscribe_url(self, cr, uid, mail, email_to, msg=None, context=None):
                 'params': urllib.urlencode({'db': cr.dbname, 'res_id': mail.res_id, 'email': email_to})
             }
         )
-        return '<small><a href="%s">%s</a></small>' % (url, msg or 'Click to unsubscribe')
+        return url, msg
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mail_mail.py]: Why changing the return style of this one, especially that there are other methods always returning html content in the mail post processing ?

File: addons/mass_mailing/models/mass_mailing.py [hunk 3]
```diff
@@ -809,6 +812,8 @@ def convert_link(self, cr, uid, ids, context=None):
             res[mass_mailing.id] = mass_mailing.body_html if mass_mailing.body_html else ''
             
             for match in re.findall(URL_REGEX, res[mass_mailing.id]):
+                if '/unsubscribe_from_list' in match[1]:
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mass_mailing.py]: Please avoid doing "do something unless some custom stuff".

File: addons/mass_mailing/models/mass_mailing.py [hunk 4]
```diff
@@ -868,6 +873,8 @@ def send_get_mail_body(self, mail, partner=None):
         """Override to add Statistic_id in shorted urls """
         if mail.mailing_id and mail.body_html:
             for match in re.findall(URL_REGEX, mail.body_html):
+                if '/unsubscribe_from_list' in match[1]:
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mass_mailing.py]: Please avoid doing "do something unless some custom not documented stuff".

File: addons/mass_mailing/wizard/test_mailing.py [hunk 5]
```diff
@@ -35,8 +35,6 @@ def send_mail_test(self, cr, uid, ids, context=None):
                     'mailing_id': mailing.id,
                 }
                 mail_mail_obj = Mail.browse(cr, uid, Mail.create(cr, uid, mail_values, context=context), context=context)
-                unsubscribe_url = Mail._get_unsubscribe_url(cr, uid, mail_mail_obj, test_mail, context=context)
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/wizard/test_mailing.py]: Why this change ?

File: addons/website_mail/models/email_template.py [hunk 6]
```diff
@@ -1,19 +1,23 @@
...
 
-class EmailTemplate(osv.Model):
+class EmailTemplate(models.Model):
     _inherit = 'mail.template'
 
-    def action_edit_html(self, cr, uid, ids, context=None):
-        if not len(ids) == 1:
-            raise ValueError('One and only one ID allowed for this action')
-        url = '/website_mail/email_designer?model=mail.template&res_id=%d&enable_editor=1' % (ids[0],)
+    theme_xml_id = fields.Char('Theme XML ID',
```
Comment (tde-banana-odoo) [inline addons/website_mail/models/email_template.py]: What is the purpose of this field ? A lot of templates are used for other purpose than mass mailing through web designer, what will be the impact ?

**Comment by Gorash** [comment]
I'll take care to change and adapt this branch for the new backend editor (don't merge directly)
@tde-banana-odoo: Thanks for the comments and review it will facilitate my work

File: addons/mass_mailing/models/mail_mail.py (hunk 1 above)
Comment (dja-odoo) [inline addons/mass_mailing/models/mail_mail.py]: Because we are giving Link to unsubscribe In each footer snippet of all the themes so we have to set the proper href there. And we can't have unsubscribe link at the time of designing. It will be generated while sending the mails (link needs recordset  of created mail that we won't be having while the time of designing).

File: addons/mass_mailing/models/mail_mail.py (hunk 2 above)
Comment (dja-odoo) [inline addons/mass_mailing/models/mail_mail.py]: Previously we were manually appending the the un-subscription link after mail body with help of this method. But now, as asked in specification, we will already have html content for the link in mail body and we will just set the link there, so we won't need to return html content now.

File: addons/mass_mailing/models/mass_mailing.py (hunk 3 above)
Comment (dja-odoo) [inline addons/mass_mailing/models/mass_mailing.py]: If this link is also converted into shortened URL, there won't be any way to find it in mail body and so we won't be able to replace the link at appropriate place while sending mail.

File: addons/mass_mailing/models/mass_mailing.py (hunk 4 above)
Comment (dja-odoo) [inline addons/mass_mailing/models/mass_mailing.py]: Same as I explained.

File: addons/mass_mailing/wizard/test_mailing.py (hunk 5 above)
Comment (dja-odoo) [inline addons/mass_mailing/wizard/test_mailing.py]: Because we will have Un-subscription link already in mail body now. Also in the previous code, there was a line 
body = tools.append_content_to_html(mailing.body_html, unsubscribe_url, plaintext=False, container_tag='p') . 
But "body" variable wasn't used any-where and that link was never been sent actually in test mail.

File: addons/website_mail/models/email_template.py (hunk 6 above)
Comment (dja-odoo) [inline addons/website_mail/models/email_template.py]: It stores the last selected theme used for designing that particular mail template. If user starts designing the same template again, last theme used for designing it will be loaded automatically instead of asking to select the one.

File: addons/website/static/src/js/website.tour.rte.js [hunk 7]
```diff
@@ -61,12 +61,12 @@
             },
             {
                 waitFor:   '#wrapwrap > main > div > section .row > div:first',
-                element:   '.note-popover button[data-event="color"]',
+                element:   '.note-air-popover button[data-event="color"]',
```
Comment (Gorash) [inline addons/website/static/src/js/website.tour.rte.js]: Unauthorized to change website_rte
The rte is not supposed to have a different behavior for this task.

File: addons/website/static/lib/summernote/src/js/Renderer.js [hunk 8]
```diff
@@ -409,7 +409,8 @@ define([
         var content = '<div class="btn-group">' + fullButton + halfButton + quarterButton + '</div>' +
                       '<div class="btn-group">' + leftButton + rightButton + justifyButton + '</div>' +
                       '<div class="btn-group">' + roundedButton + circleButton + thumbnailButton + noneButton + '</div>' +
-                      '<div class="btn-group">' + removeButton + '</div>';
+                      '<div class="btn-group">' + removeButton + '</div>'+
+                      '<div class="note-color btn-group hidden" style="display:none;">'+tplButtonInfo.color(lang, options) + '</div>';
```
Comment (Gorash) [inline addons/website/static/lib/summernote/src/js/Renderer.js]: Unauthorized to change summernote
Otherwise I could not follow up and work with the official library
